objects/src/notes/note_execution_hint.rs

```diff
@@ -0,0 +1,182 @@
+    None,
+    /// The note's script can be executed at any time.
+    Always,
+    /// The note's script can be executed after the specified block height.
+    AfterBlock { block_num: u32 },
+    /// The note's script can be executed in the specified slot within the specified epoch.
+    ///
+    /// The slot is defined as follows:
+    /// - First we define the length of the epoch in powers of 2. For example, epoch_len = 10 is an epoch
+    ///   of 1024 blocks.
+    /// - Then we define the length of a slot within the epoch also using powers of 2. For example,
+    ///   slot_len = 7 is a slot of 128 blocks.
+    /// - Lastly, the offset specifies the index of the slot within the epoch - i.e., 0 is the first slot,
+    ///   1 is the second slot etc.
+    ///
+    /// For example: { epoch_len: 10, slot_len: 7, slot_offset: 1 } means that the note can
+    /// be executed in any second 128 block slot of a 1024 block epoch. These would be blocks 128..255,
+    /// 1152..1279, 2176..2303 etc.
+    OnBlockSlot {
+        epoch_len: u8,
+        slot_len: u8,
+        slot_offset: u8,
+    },
+}
+
+impl NoteExecutionHint {
+    /// Creates a [NoteExecutionHint::None] variant
+    pub fn none() -> Self {
+        NoteExecutionHint::None
+    }
+
+    /// Creates a [NoteExecutionHint::Always] variant
+    pub fn always() -> Self {
+        NoteExecutionHint::Always
+    }
+
+    /// Creates a [NoteExecutionHint::AfterBlock] variant based on the given `block_num`
+    pub fn after_block(block_num: u32) -> Self {
+        NoteExecutionHint::AfterBlock { block_num }
+    }
+
+    /// Creates a [NoteExecutionHint::OnBlockSlot] for the given parameters
+    pub fn on_block_slot(epoch_len: u8, slot_len: u8, slot_offset: u8) -> Self {
+        NoteExecutionHint::OnBlockSlot { epoch_len, slot_len, slot_offset }
+    }
+
+    /// Returns whether the note execution conditions validate for the given `block_num`
+    pub fn can_be_consumed(&self, block_num: u32) -> bool {
+        match self {
+            NoteExecutionHint::None => false,
+            NoteExecutionHint::Always => true,
+            NoteExecutionHint::AfterBlock { block_num: hint_block_num } => {
+                block_num >= *hint_block_num
+            },
+            NoteExecutionHint::OnBlockSlot { epoch_len, slot_len, slot_offset } => {
+                let epoch_len_blocks: u32 = 1 << epoch_len;
+                let slot_len_blocks: u32 = 1 << slot_len;
+
+                let block_epoch_index = block_num / epoch_len_blocks;
+
+                let slot_start_block =
+                    block_epoch_index * epoch_len_blocks + (*slot_offset as u32) * slot_len_blocks;
+                let slot_end_block = slot_start_block + slot_len_blocks;
+
+                block_num >= slot_start_block && block_num < slot_end_block
+            },
+        }
+    }
+}
+
+// CONVERSIONS
+// ================================================================================================
```

Comment:
Depending on how we encode the hint into metadata, some of this conversion may or may not be relevant. For example, it may make sense to provide something like `from_parts(tag, payload)` and `into_parts() -> (u8, u32)` methods instead.